### PR TITLE
udev: add basic set USDTs

### PR DESCRIPTION
**Pull request by msekletar**
I am submitting this PR as an initial Draft to get first round of comments on the list of trace-points. I tried to pick such tracepoints which would allow people debug very specific issues on production systems. 

Using combination of these tracepoints it should be possible answer questions like,
- What are all the rules executed for second change event on /dev/dm-0?
- How many worker processes we have spawned during boot on system with 1000 disk drives?
- How many external helpers we executed during the boot on such system?
- Were there any timeouts during the event processing?

It should be fairly easy to translate these questions to bpftrace [1] scripts and hence got answers for them also on production systems (without debuginfo and debuggers installed) all while not disrupting/slowing the boot with full blown debug logging or installing debug builds on production boxes.

[1] https://github.com/iovisor/bpftrace

### CODE REVIEW

File: src/udev/meson.build
Comment (mrc0mmand) [inline src/udev/meson.build]:
```suggestion
        libudevd_core_sources += [probes_h]
```

File: src/udev/meson.build
Comment (poettering) [inline src/udev/meson.build]: spurious double empty line

**Review by poettering — COMMENTED** [review]
any chance you can add to the commit msg a very brief explanation how to actually use this?

hmm is "stap" systemtap? Is that a thing outside of rhel? And the reference to dtrace, isn't that an orace specific thing? Can you enlighten the uneducated reader (like me) very briefly about this in the commit msg or so?

File: src/basic/macro.h
Comment (poettering) [inline src/basic/macro.h]: c compiler says double underscores are owned by compiler. maybe just use a single one here, should be enough too

File: src/udev/udev-event.c
Comment (poettering) [inline src/udev/udev-event.c]: hmm, which package provides this header? i don't have it here?

File: src/udev/probes.d
Comment (poettering) [inline src/udev/probes.d]: hmm, shouldn't thee comments be inside the parameters list? i.e. is that really how this should be done that you put the args nameless in there, and add the arg names only as comment?

File: src/udev/probes.d
Comment (poettering) [inline src/udev/probes.d]: hmm, they use ".d" as suffix for a file? how unfortunate since that's usually used for drop-in style dirs

File: src/udev/meson.build
Comment (poettering) [inline src/udev/meson.build]: you are basically building a static lib with this, right? isn't the right approach for that to use meson's native static lib concepts? (i have no idea, just asking)

File: src/basic/macro.h
Comment (poettering) [inline src/basic/macro.h]: to make this double eval free I think it would be better to copy the device ptr to a local variable too, and then use that for all calls, i.e. add `sd_device *_d = dev;` to the top of the block

File: src/basic/macro.h
Comment (poettering) [inline src/basic/macro.h]: double newline

**Review by poettering — COMMENTED** [review]
looks great btw, but didn't test this, don't know how ;-)

File: src/udev/udev-event.c
Comment (poettering) [reply src/udev/udev-event.c]: needs some ifdeffery i guess

File: src/basic/macro.h
Comment (poettering) [inline src/basic/macro.h]: hmm, given this is sd_device specific this should really be in some sd-device header, and be named DEVICE_TRACE_POINT() or so

File: meson.build
Comment (poettering) [inline meson.build]: `have` i not set to false if stap isn't enabled

File: src/udev/meson.build
Comment (keszybz) [inline src/udev/meson.build]: Maybe check if meson generators can be used for this? (We didn't use them so far because in the past they didn't work too well, but meson has progressed quite a bit, so maybe it'd be worth trying again.)

File: src/udev/meson.build
Comment (keszybz) [reply src/udev/meson.build]: Yeah, this should be done using `static_library`.

File: src/udev/meson.build
Comment (keszybz) [inline src/udev/meson.build]: [] not necessary.

File: src/udev/probes.d
Comment (keszybz) [inline src/udev/probes.d]: Shouldn't this be autogenerated from annotations in the source file? Keeping this list will be a bother if we have to do it by hand.

File: src/udev/meson.build
Comment (msekletar) [reply src/udev/meson.build]: `static_library()` requires as a second argument list of sources to build library from. However, I need to create static library from already prebuilt object file.

File: src/udev/meson.build
Comment (msekletar) [reply src/udev/meson.build]: I'd need to use two generators where output of one is input of the other. That unfortunately still doesn't work with current meson,

https://github.com/mesonbuild/meson/issues/1141
https://github.com/mesonbuild/meson/issues/5401

File: src/udev/probes.d
Comment (msekletar) [reply src/udev/probes.d]: They used that suffix waaay before we came up with idea of drop-in directories.

File: src/udev/probes.d
Comment (msekletar) [reply src/udev/probes.d]: Well, this is actually the canonical source. We actually generate the source file that defines the implementation of the each tracepoint from this file.

Also, having to maintain the list in two places is actually a good thing. You don't want to accidentally change your trace point format. Think of this in similar terms like linker `.sym` files.

File: src/udev/udev-event.c
Comment (msekletar) [reply src/udev/udev-event.c]: Good point, these needs to be ifdefed. Btw, it is systemtap-devel.

File: src/udev/probes.d
Comment (msekletar) [reply src/udev/probes.d]: Well, I based this of off "official" documentation.

https://sourceware.org/systemtap/wiki/AddingUserSpaceProbingToApps

**Comment by msekletar** [comment]
I fixed most of the issues pointed out so far. I've also expanded the commit message and included some examples. I think that this is now in good enough shape to move out of Draft state.

@poettering @keszybz PTAL.

**Comment by poettering** [comment]
hmm, needs rebase

File: src/shared/udev-util.h
Comment (poettering) [inline src/shared/udev-util.h]: doesn't matter, but spurious space before the =

File: src/shared/udev-util.h
Comment (poettering) [inline src/shared/udev-util.h]: missing newline at end of file

File: src/udev/udev-event.c
Comment (poettering) [inline src/udev/udev-event.c]: hmm, our coding style usually mandates that `<>` includes happen before `""` includes. I'd prefer if the same would be done here, even if thta means we'll have two `#if` pairs

File: src/udev/udev-rules.c
Comment (poettering) [inline src/udev/udev-rules.c]: ditto

File: src/udev/udev-event.c
Comment (poettering) [inline src/udev/udev-event.c]: hmm, thinking about this, wouldn't it be better, to move this into udev-util.h, given that that' where DEVICE_TRACE_POINT is defined, too?

Comment (poettering) [reply src/udev/udev-event.c]: or taking this idea further, maybe add udev-trace-point.h for this? the macros is complex enough to deserve its own header file i think.

File: src/shared/udev-util.h
Comment (poettering) [inline src/shared/udev-util.h]: hmm, I figure this should save/restore errno, so that it is without side-effect. i.e. maybe add PROTECT_ERRNO to the beginning of the do loop.

compare with the log functions where we always do that.

**Review by poettering — CHANGES_REQUESTED** [review]
looks excellent, some minor stuff.

I'll leave the review of the meson stuff to @keszybz 

Once we landed this I'd love to see more of these added, including to PID 1

**Comment by poettering** [comment]
(Dropped the "RFC" from the PR title, since I think we all agree this is totally great, and it just needs some nitpicks fixed before merge)

File: src/udev/udev-event.c
Comment (t-8ch) [reply src/udev/udev-event.c]: Given that this file does not use anything from `sys/sdt.h` and the generated header file already includes `sys/sdt.h` it can probably be removed completely.

**Comment by t-8ch** [comment]
Out of curiosity: What is the advantage of using the generated header files over directly using `STAP_PROBE()`/`DTRACE_PROBE()`?

**Comment by poettering** [comment]
@msekletar what's your take on this? merge @t-8ch's work into this?

i sympathize with having fewer deps...

**Comment by msekletar** [comment]
Sure, I am fine with not using dtrace tool. @t-8ch with you permission I will apply your patch on top of my branch (after I do cleanups pointed out above) and resubmit it here. Is that ok?

**Comment by msekletar** [comment]
Ahh, looking at the list of things that I was supposed to fix it seems most the are not relevant after applying patch from @t-8ch

**Comment by t-8ch** [comment]
> I will apply your patch on top of my branch

@msekletar Sure!

**Comment by msekletar** [comment]
I've applied patch from @t-8ch on top off my branch and rebased the entire thing. @poettering PTAL

File: src/udev/meson.build
Comment (msekletar) [reply src/udev/meson.build]: Not relevant any longer.

File: src/udev/meson.build
Comment (msekletar) [reply src/udev/meson.build]: Same here.

File: src/udev/meson.build
Comment (msekletar) [reply src/udev/meson.build]: Again, not relevant after changes after @t-8ch.

File: src/shared/udev-util.h
Comment (msekletar) [reply src/shared/udev-util.h]: PROTECT_ERRNO added by @t-8ch

File: src/udev/udev-event.c
Comment (msekletar) [reply src/udev/udev-event.c]: Yeah we may want to add separate file in the future. Let's stick to `udev-util.h` for now.

File: src/shared/udev-util.h
Comment (t-8ch) [inline src/shared/udev-util.h]: This `args` should actually be removed. It's a leftover from an earlier revision.

File: src/shared/udev-util.h
Comment (dnicolodi) [inline src/shared/udev-util.h]: While you are at it: something went wrong with the placement of the line continuation `\` here. Maybe a tabs vs spaces problem?

**Comment by poettering** [comment]
Hmm, the second commit undoes half of what the first commit added. Can we please squash them into one? It's weird reviewing stuff that is immediately deleted within the same PR again.

Yes, this makes correct attribution hard, but maybe just say in the commit msg itself who contributed to this.

File: src/shared/udev-util.h
Comment (poettering) [inline src/shared/udev-util.h]: hmm, i figure we also want to include the sysfs path and the subsystem here.

**Comment by poettering** [comment]
I review the overall diff, looks excellent. I like the simplicity of the resulting patch a lot.  CI isn't quite as happy as I am though ;-)

File: src/shared/udev-util.h
Comment (msekletar) [reply src/shared/udev-util.h]: Good point!

**Comment by t-8ch** [comment]
The CI failure is due to https://github.com/systemd/systemd/pull/19015#discussion_r649786199 where I unfortunately forgot to clean up the macro definition.

File: src/udev/udev-rules.c
Comment (t-8ch) [inline src/udev/udev-rules.c]: This is not needed anymore.

File: src/udev/udev-rules.c
Comment (t-8ch) [inline src/udev/udev-rules.c:32]: This is also included transitively so this could in theory be dropped.

Comment (msekletar) [reply src/udev/udev-rules.c:32]: I think we shouldn't be relying on transitivity of includes and always be explicit.

File: src/shared/udev-util.h
Comment (msekletar) [reply src/shared/udev-util.h]: Ok, I've added those

File: src/shared/udev-util.h
Comment (yuwata) [inline src/shared/udev-util.h]: Maybe, it is safer to add `()` around `dev`: `sd_device *_d = (dev);`.

Comment (msekletar) [reply src/shared/udev-util.h]: Fixed

**Comment by msekletar** [comment]
This is now ready for re-review. @poettering PTAL.

**Comment by yuwata** [comment]
Superficially LGTM. But I am not familiar with that feature. Let's wait for another reviewer.

**Comment by poettering** [comment]
lgtm.

might be nice to extend the commit msg, and mentioning @t-8ch's work this incorporates, no?

**Comment by yuwata** [comment]
> might be nice to extend the commit msg, and mentioning @t-8ch's work this incorporates, no?

It is already mentioned.

**Comment by keszybz** [comment]
LGTM too.
CI failures look unrelated, let's merge.

**Pull request merged**
